### PR TITLE
Creating a Hierarchy of Animal Classes

### PR DESCRIPTION
**Pull request by Vladimir2075**

### CODE REVIEW

File: src/Main/Island/Organism/Animals/Animal.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import Main.Island.Organism.Organism;
+import Main.Island.СharacteristicsOfAnimals;
+
+import java.util.HashMap;
+
+public abstract class Animal extends Organism {
+    private float amountFoodPerIteration;
+    private int maxSpeed;
+    private static final int ZERO_PROBABILITIES = 0;
+    private float lifeLevel = 100f;
```
Comment (Volodymyr95) [inline src/Main/Island/Organism/Animals/Animal.java]: why float?

File: src/Main/Island/Organism/Animals/Herbivorous/Buffalo.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package Main.Island.Organism.Animals.Herbivorous;
+public class Buffalo extends HerbivorAnimal {
+    String  logo = "Buffalo ";
```
Comment (Volodymyr95) [inline src/Main/Island/Organism/Animals/Herbivorous/Buffalo.java]: use icon code

File: src/Main/Island/Organism/Animals/Animal.java (hunk 1 above)
Comment (Vladimir2075) [reply src/Main/Island/Organism/Animals/Animal.java]: тому що є не цілочисленні значення

File: src/Main/Island/Organism/Animals/Herbivorous/Buffalo.java (hunk 2 above)
Comment (Vladimir2075) [reply src/Main/Island/Organism/Animals/Herbivorous/Buffalo.java]: не знайшов відповідної ікони

File: src/Main/java/Island/Organism/Animals/Animal.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    private static final int UP_DIRECTION = 0;
+    private static final int DOWN_DIRECTION = 1;
+    private static final int LEFT_DIRECTION = 2;
+    private static final int Max_COUNT_ATTEMPT = 5;
+    private MapIsland mapIsland = MapIsland.getInstance();
+    private float amountFoodPerIteration;
+    private int maxSpeed;
+    private float lifeLevel = 100f;
+    private float lossEnergyLife = 20;
+    private boolean isAlive =true;
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Animal.java]: formatting: add empty line between

File: src/Main/java/Island/Organism/Animals/Animal.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        Cell currentField = this.getFieldOrganism();
+        int countRelatedAnimal = currentField.getNumberOfLivingOrganisms().get(this.getClass().getSimpleName());
+        if ((countRelatedAnimal > 1) && Cell.getMaxCountOrganismPerCell(this.getClass().getSimpleName())>countRelatedAnimal) {
+            try {
+                Organism organism = (Organism) this.getClass().getDeclaredConstructor().newInstance();
+                organism.setFieldOrganism(this.getFieldOrganism());
+                this.getFieldOrganism().addToOrganisms(organism);
+                mapIsland.birthOfOrganism(organism);
+            } catch (InstantiationException e) {
+                throw new RuntimeException(e);
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Animal.java:113]: add informative exception message

File: src/Main/java/Island/Organism/Animals/Animal.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+                        this.setFieldOrganism(newField);
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+        }
+    }
+    public void eat (){
+        //this.
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Animal.java]: delete it

File: src/Main/java/Island/Cell.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
+package Island;
```
Comment (Volodymyr95) [inline src/Main/java/Island/Cell.java:1]: apply formatting for all classes

File: src/Main/java/Island/Cell.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+import Island.Organism.Organism;
+import java.util.*;
+public class Cell {
```
Comment (Volodymyr95) [inline src/Main/java/Island/Cell.java:5]: formatting

File: src/Main/java/Island/LifeOnTheIsland.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    public void runLifeOnTheIsland () {
+        new СreatingIslandWithLivingCreatures().initialisationIsland();
+        Statistics statistics = Statistics.getInstance();
+        ScheduledExecutorService executorServiceStatistic = Executors.newScheduledThreadPool(1);
+        executorServiceStatistic.scheduleWithFixedDelay(statistics, 0, map.getDurationSimulationCycle(), TimeUnit.MILLISECONDS);
+        try {
+            Thread.sleep(5000);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
```
Comment (Volodymyr95) [inline src/Main/java/Island/LifeOnTheIsland.java:42]: add informative message

File: src/Main/java/Island/MapIsland.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    public void DeathOfOrganism (Organism organism) {
+        HashMap<Class,Integer> map;
+        if (organism instanceof PredatorAnimal ) {
+            map = statistics.getStatisticsOfAnimalPredators();
+        } else if (organism instanceof HerbivorAnimal) {
+            map = statistics.getStatisticsOfAnimalHerbivorous();
+        } else {
+            map = statistics.getStatisticsOfPlants();
+         //       if (listOfPlants.indexOf(organism)>0) {
```
Comment (Volodymyr95) [inline src/Main/java/Island/MapIsland.java]: delete commented code

File: src/Main/java/Island/MapIsland.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+         //           this.listOfPlants.remove(organism);
+         //       }
+        }
+        synchronized (map) {
+            if (map.get(organism.getClass())>1) {
```
Comment (Volodymyr95) [inline src/Main/java/Island/MapIsland.java]: what 1 mean?

File: src/Main/java/Island/MapIsland.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            }
+        }
+        HashMap<Class,Integer> mapDeathOrganism =statistics.getStatisticsOfDeath();
+            if (mapDeathOrganism.get(organism.getClass()) == null) {
+                mapDeathOrganism.put(organism.getClass(), 1);
+            } else {
+                mapDeathOrganism.put(organism.getClass(), mapDeathOrganism.get(organism.getClass()) + 1);
+            }
+    }
+    private MapIsland() {
```
Comment (Volodymyr95) [inline src/Main/java/Island/MapIsland.java]: why constructor after methods?

File: src/Main/java/Island/Organism/Animals/Animal.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    }
+    private void setAmountFoodPerIteration() {
+        this.amountFoodPerIteration = characteristicsOfAnimals.getAmountFoodPerIterationOfTheAnimals().get(this.getClass().getSimpleName());
+    }
+    public Thread getCurrentTread() {
+        return currentTread;
+    }
+    public int getProbabilityKillingAnimal(String classFood) {
+        String classKiller = this.getClass().getSimpleName();
+        Integer res;
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Animal.java:78]: not informative variable name

File: src/Main/java/Island/Organism/Animals/Animal.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        HashMap<String, Integer> killerFood = (HashMap<String, Integer>) characteristicsOfAnimals.getEatAnimalProbabilityMatrix().get(classKiller);
+        if (killerFood != null) {
+            res = killerFood.get(classFood);
+        } else {
+            res = ZERO_PROBABILITIES;
+        }
+        if (res == null) {
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Animal.java:85]: Objects.isNull()

File: src/Main/java/Island/Organism/Animals/Animal.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+                if (statistics.getStatisticsOfBirth().get(organism.getClass()) == null) {
+                    statistics.getStatisticsOfBirth().put(organism.getClass(), MIN_COUNT_ORGANISM);
+                } else {
+                    statistics.getStatisticsOfBirth().put(organism.getClass(), statistics.getStatisticsOfBirth().get(organism.getClass()) + MIN_COUNT_ORGANISM);
+                }
+
+            } catch (InstantiationException e) {
+                throw new RuntimeException(e);
+            } catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Animal.java:115]: where exception messages?

File: src/Main/java/Island/Organism/Animals/Herbivorous/Boar.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+                        case 1: if (this.isAlive()) {this.move();break;}
+                        case 2: if (this.isAlive()) {this.multiply();break;}
+                        case 3: if (this.isAlive()) {this.eat();break;}
+                        case 4: if (this.isAlive()) { this.move(); break; }
+                    }
+                }
+                Thread.sleep(MapIsland.TIME_WAIT_BETWEEN_ACTIONS_ANIMAL);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Herbivorous/Boar.java]: don't catch Exception, use specific class for that

File: src/Main/java/Island/Organism/Animals/Herbivorous/Buffalo.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        return LOGO;
+    }
+    @Override
+    public void run() {
+        this.setRunThread(true);
+        while (this.isAlive()) {
+            int index = RANDOM.nextInt(COUNT_ACTIONS_ANIMAL);
+            try {
+                for (int i = 0; i < COUNT_ACTIONS_ANIMAL; i++) {
+                    switch (i + index) {
```
Comment (Volodymyr95) [inline src/Main/java/Island/Organism/Animals/Herbivorous/Buffalo.java]: u have same code in 2 different classes, move it and reuse

File: pom.xml [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>groupId</groupId>
+    <artifactId>Wild_Island</artifactId>
+    <version>1.0-SNAPSHOT</version>
+
+    <properties>
+        <maven.compiler.source>19</maven.compiler.source>
```
Comment (Volodymyr95) [inline pom.xml:12]: why 19?

File: src/Main/java/Island/Cell.java [hunk 18]
```diff
@@ -0,0 +1,94 @@
+package Island;
+import Island.Organism.Organism;
+import java.util.*;
+public class Cell {
+    private static final Integer ZER0_COUNT =0;
+    private static final Integer MIN_COUNT_ANIMAL = 1;
+    private volatile List<Organism> organisms = new LinkedList<>();
+    private int X;
```
Comment (Volodymyr95) [inline src/Main/java/Island/Cell.java]: should be x not X

File: src/Main/java/Island/Cell.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    private int Y;
```
Comment (Volodymyr95) [inline src/Main/java/Island/Cell.java]: should be y, not Y

File: src/Main/java/Island/Cell.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
+package Island;
+import Island.Organism.Organism;
+import java.util.*;
+public class Cell {
+    private static final Integer ZER0_COUNT =0;
```
Comment (Volodymyr95) [inline src/Main/java/Island/Cell.java]: formatting

File: src/Main/java/Island/Cell.java [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+            } else {
+                numberOfLivingOrganisms.replace(organism.getClass().getSimpleName(), varCount, varCount-MIN_COUNT_ANIMAL);
+            }
+        }
+        this.organisms.remove(organism);
+    }
+    public List<Organism> getOrganisms() {
+        return organisms;
+    }
+    public Cell(int x, int y) {
```
Comment (Volodymyr95) [inline src/Main/java/Island/Cell.java]: constructor should be before methods

File: src/Main/java/Island/LifeOnTheIsland.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+                if (animal instanceof PredatorAnimal) {
+                    if (animal.isAlive()) {
+                        return false;
+                    }
+                }
+            }
+        }
+        return true;
+    }
+    class IslandUncaughtExceptionHandler implements Thread.UncaughtExceptionHandler {
```
Comment (Volodymyr95) [inline src/Main/java/Island/LifeOnTheIsland.java:21]: make it lambda

File: src/Main/java/Island/LifeOnTheIsland.java [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+        }
+    }
+    class HandlerThreadFactory implements ThreadFactory {
```
Comment (Volodymyr95) [inline src/Main/java/Island/LifeOnTheIsland.java:26]: make it lambda

File: src/Main/java/Island/LifeOnTheIsland.java (hunk 8 above)
Comment (Volodymyr95) [reply src/Main/java/Island/LifeOnTheIsland.java:42]: not resolved

File: src/Main/java/Island/LifeOnTheIsland.java [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+                if (checkExitConditional()) {break;}
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        executorServiceAnimalLifeCycle.close();
+        try {
+            Thread.sleep(map.getDurationSimulationCycle());
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
```
Comment (Volodymyr95) [inline src/Main/java/Island/LifeOnTheIsland.java:62]: not informative message

File: src/Main/java/Island/Main.java [hunk 25]
```diff
@@ -0,0 +1,7 @@
+package Island;
+
+public class Main {
+    public static void main(String[] args) {
+              new LifeOnTheIsland().runLifeOnTheIsland();
```
Comment (Volodymyr95) [inline src/Main/java/Island/Main.java:5]: new LifeOnTheIsland().run();

File: src/Main/java/Island/MapIsland.java [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+            for (Animal animal : listOfAnimal) {
+                if (!animal.isRunThread()) {
+                    result.add(animal);
+                }
+            }
+        }
+        return result;
+    }
+    public List<Plant> getListOfPlants(Cell cell)
+    {
```
Comment (Volodymyr95) [inline src/Main/java/Island/MapIsland.java:63]: formatting

**Review by Volodymyr95 — APPROVED** [review]

**Pull request merged**